pkg/controller/appdefinition/acorn.go

```diff
@@ -77,19 +78,29 @@ func scopeLinks(app *v1.AppInstance, bindings v1.ServiceBindings) (result v1.Ser
 }
 
 func toAcorn(appInstance *v1.AppInstance, tag name.Reference, pullSecrets *PullSecrets, acornName string, acorn v1.Acorn) *v1.AppInstance {
-	image := images.ResolveTag(tag, acorn.Image)
-	if strings.HasPrefix(acorn.Image, "sha256:") {
-		image = strings.TrimPrefix(acorn.Image, "sha256:")
+	var image string
+	_, isPattern := autoupgrade.AutoUpgradePattern(acorn.Image)
+	isPattern = isPattern && acorn.AutoUpgrade != nil && *acorn.AutoUpgrade
+	if isPattern {
+		image = acorn.Image
+
+		// remove the autoupgrade pattern from the end of the image for resolving the pull secret
+		// the registry is all that really matters for the pull secret so this is safe to do
+		pullSecrets.ForAcorn(acornName, image[:strings.LastIndex(image, ":")])
```

Comment:
nit: if you don't ignore the first return of the call to `autoupgrade.AutoUpgradePattern` above (say you call it `pattern`), then this could be
```suggestion
		pullSecrets.ForAcorn(acornName, strings.TrimSuffix(image, pattern))
```

Reply:
Good point, that is a lot cleaner. Updated.